internal/tasks/account_deletion.go

```diff
@@ -0,0 +1,213 @@
+	"github.com/sapcc/go-api-declarations/cadf"
+	"github.com/sapcc/go-bits/audittools"
+	"github.com/sapcc/go-bits/jobloop"
+	"github.com/sapcc/go-bits/logg"
+	"github.com/sapcc/go-bits/sqlext"
+
+	"github.com/sapcc/keppel/internal/keppel"
+	"github.com/sapcc/keppel/internal/models"
+	"github.com/sapcc/keppel/internal/processor"
+)
+
+// EnforceManagedAccounts is a job. Each task creates newly discovered accounts from the driver.
+func (j *Janitor) DeleteAccountsJob(registerer prometheus.Registerer) jobloop.Job {
+	return (&jobloop.ProducerConsumerJob[models.AccountName]{
+		Metadata: jobloop.JobMetadata{
+			ReadableName: "delete accounts marked for deletion",
+			CounterOpts: prometheus.CounterOpts{
+				Name: "keppel_account_deletion",
+				Help: "Counter for deleted accounts.",
+			},
+		},
+		DiscoverTask: j.discoverAccountForDeletion,
+		ProcessTask:  j.deleteMarkedAccount,
+	}).Setup(registerer)
+}
+
+var (
+	accountDeletionSelectQuery = sqlext.SimplifyWhitespace(`
+		SELECT name FROM accounts
+		WHERE is_deleting AND next_deletion_attempt_at < $1
+		ORDER BY next_deletion_attempt_at ASC, name ASC
+	`)
+)
+
+func (j *Janitor) discoverAccountForDeletion(_ context.Context, _ prometheus.Labels) (accountName models.AccountName, err error) {
+	err = j.db.SelectOne(&accountName, accountDeletionSelectQuery, j.timeNow())
+	return accountName, err
+}
+
+func (j *Janitor) deleteMarkedAccount(ctx context.Context, accountName models.AccountName, labels prometheus.Labels) error {
+	accountModel, err := keppel.FindAccount(j.db, accountName)
+	if err != nil {
+		return err
+	}
+	if errors.Is(err, sql.ErrNoRows) {
+		// assume the account got already deleted
+		return nil
+	}
+
+	actx := keppel.AuditContext{
+		UserIdentity: janitorUserIdentity{TaskName: "account-deletion"},
+		Request:      janitorDummyRequest,
+	}
+	err = j.DeleteAccount(ctx, *accountModel, actx)
+	if err != nil {
+		return fmt.Errorf("could not delete account %q: %w", accountName, err)
+	}
+	return nil
+}
+
+var (
+	deleteAccountFindManifestsQuery = sqlext.SimplifyWhitespace(`
+		SELECT r.name, m.digest
+			FROM manifests m
+			JOIN repos r ON m.repo_id = r.id
+			JOIN accounts a ON a.name = r.account_name
+			LEFT OUTER JOIN manifest_manifest_refs mmr ON mmr.repo_id = r.id AND m.digest = mmr.child_digest
+    WHERE a.name = $1 AND parent_digest IS NULL
+	`)
+	deleteAccountReposQuery                   = `DELETE FROM repos WHERE account_name = $1`
+	deleteAccountCountBlobsQuery              = `SELECT COUNT(id) FROM blobs WHERE account_name = $1`
+	deleteAccountScheduleBlobSweepQuery       = `UPDATE accounts SET next_blob_sweep_at = $2 WHERE name = $1`
+	deleteAccountMarkAllBlobsForDeletionQuery = `UPDATE blobs SET can_be_deleted_at = $2 WHERE account_name = $1`
+)
+
+type deleteAccountRemainingManifest struct {
+	RepositoryName string
+	Digest         string
+}
+
+func (j *Janitor) DeleteAccount(ctx context.Context, account models.Account, actx keppel.AuditContext) error {
```

Comment:
This can be a private method. In fact, I would just inline it with `deleteMarkedAccount` since that's the only callsite.